File: KisaragiTech.Dape/User/Service/PasswordGenerator.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace KisaragiTech.Dape.User.Service;

public static class PasswordGenerator
{
    // 文字集合（視認性も考慮してl,1,O,0は除外）
    private const string Alphabet =
        // 大文字 (O除く)
        "ABCDEFGHJKLMNPQRSTUVWXYZ" +
        // 小文字 (l除く)
        "abcdefghijkmnopqrstuvwxyz" +
        // 数字 (0と1除く)
        "23456789" +
        // 記号（調整可）
        "!@#$%^&*()-_=+[]{}|;:,.<>?";

    public static string Generate(int length = 20)
    {
        if (length <= 0)
        {
            throw new ArgumentException("length must be positive", nameof(length));
        }

        var sb = new StringBuilder(length);

        for (var i = 0; i < length; i++)
        {
            var index = RandomNumberGenerator.GetInt32(Alphabet.Length);
            sb.Append(Alphabet[index]);
        }

        return sb.ToString();
    }
}

File: src/main/scala/io/github/ksrgtech/dape/user/service/PasswordGenerator.scala
package io.github.ksrgtech.dape.user.service

import cats.MonadThrow
import cats.effect.std.SecureRandom
import cats.implicits.given

import io.github.ksrgtech.dape.base.cats.assert

object PasswordGenerator {

  // Character set (excluding l, 1, O, 0 for readability)
  private val Alphabet: String =
    // Uppercase (excluding O)
    "ABCDEFGHJKLMNPQRSTUVWXYZ" +
      // Lowercase (excluding l)
      "abcdefghijkmnopqrstuvwxyz" +
      // Numbers (excluding 0 and 1)
      "23456789" +
      // Symbols (adjustable)
      "!@#$%^&*()-_=+[]{}|;:,.<>?"

  def generate[F[_]: {SecureRandom, MonadThrow as mt}](length: Int = 20): F[String] = for {
    _ <- mt.assert(length > 0, "length must be positive")
    sb = new StringBuilder(length)
    cc <- SecureRandom[F]
      .nextIntBounded(Alphabet.length)
      .map(Alphabet.charAt)
      .replicateA(length)
  } yield sb.appendAll(cc).toString()
}
